### PR TITLE
feat: TokenGeneration module

### PR DESCRIPTION
**Pull request by dylanjames-dd**
# feat: TokenGeneration module

This PR creates a new module - `TokenGeneration`.

This module is intended to facilitate the creation of tokens which are commonly used to secure network calls.
The first (and currently only) token type which this module supports is a [JWT (JSON Web Token)](https://datatracker.ietf.org/doc/html/rfc7519).

# Checklist

## Before raising your pull request:
- [x] Commit messages that conform to conventional commit messages
- [x] Ran the app locally ensuring it builds 
- [x] Ran the tests locally ensuring they pass on Build
- [x] Pull request has a clear title with a short description about the feature or update
- [x] Created a `draft` pull request if it is not yet ready for review

## Before your pull request can be reviewed:
- [x] Met all of the acceptance criteria specified in the user story on Jira
- [x] Reviewed your own code to ensure you are following the style guidelines
- [x] Ran the app and tested the feature on a range of device sizes
      Please include iPod Touch, iPhone SE and iPhone 11 as a minimum.
- [x] Written Unit and Integration tests if needed
- [x] Written documentation for the change

~- [ ] Met all accessibility requirements?
    - [ ] Checked dynamic type sizes are applied
    - [ ] Checked VoiceOver can navigate your new code
    - [ ] Checked a user can navigate only using a keyboard around your new code~

## Before merging your pull request:
- [x] Ensure that the code coverage and SonarCloud checks have passed
- [ ] Actioned and resolved all comments, reaching out to reviewers for clarifications if necessary.
- [ ] Ran the app to ensure that no regressions have been caused by changes during code review.

### CODE REVIEW

File: Sources/TokenGeneration/JWT/JWTGenerator.swift [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    
+    public init(jwtRepresentation: JWTRepresentation,
+                signingService: JWTSigningService) {
+        self.jwtRepresentation = jwtRepresentation
+        self.signingService = signingService
+    }
+    
+    var token: String {
+        get throws {
+            guard let headerData = try? JSONSerialization.data(withJSONObject: jwtRepresentation.header),
```
Comment (obinns-dd) [inline Sources/TokenGeneration/JWT/JWTGenerator.swift:15]: should we use `Codable` rather than `JSONSerialization` for encoding?

Comment (dylanjames-dd) [reply Sources/TokenGeneration/JWT/JWTGenerator.swift:15]: image 2024-11-19 at 16 31 47

Comment (dylanjames-dd) [reply Sources/TokenGeneration/JWT/JWTGenerator.swift:15]: Having the properties of this type serialised feels like it makes it extendable whereas having a type be decoded feels somewhat limited unless you've got a clever solution to that..? 🙂

**Review by benmckillop-dd — APPROVED** [review]

**Review by Magdaline-A — APPROVED** [review]

File: Sources/TokenGeneration/Extensions/Data+Extensions.swift [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+extension Data {
+    /// Base 64 encoding the string.
+    ///
+    /// Removing unallowed URL encoded character:
+    ///   - "="
+    ///
+    /// Replacing unallowed URL encoded characters:
+    ///   - "+" with "-"
+    ///   - "/" with "_"
+    var base64URLEncodedString: String {
```
Comment (obinns-dd) [inline Sources/TokenGeneration/Extensions/Data+Extensions.swift:12]: Do we have a link to any specifications that detail these constraints incase anyone wants further reading?

File: Sources/TokenGeneration/JWT/JWTGenerator.swift (hunk 1 above)
Comment (obinns-dd) [reply Sources/TokenGeneration/JWT/JWTGenerator.swift]: I think `any Encodable` rather than `Any` should fix that?

Comment (dylanjames-dd) [reply Sources/TokenGeneration/JWT/JWTGenerator.swift]: Gotcha, to the second point, does this limit extendable use?
If we're happy to do that I can make the change but just want to make sure we're aligned 🙂

File: Sources/TokenGeneration/Extensions/Data+Extensions.swift (hunk 2 above)
Comment (obinns-dd) [reply Sources/TokenGeneration/Extensions/Data+Extensions.swift:13]: Want to link that in the comment?

File: Sources/TokenGeneration/JWT/JWTGenerator.swift (hunk 1 above)
Comment (obinns-dd) [reply Sources/TokenGeneration/JWT/JWTGenerator.swift:15]: Cool, as per Slack, let's stick with the `JSONSerialization` as `Encodable` seems to be a pain for heterogeneous dictionaries.

**Review by obinns-dd — APPROVED** [review]

File: Sources/TokenGeneration/Extensions/Data+Extensions.swift (hunk 2 above)
Comment (robinphillips) [reply Sources/TokenGeneration/Extensions/Data+Extensions.swift:13]: I think `unallowed` should be `disallowed` or `unsupported`?

**Review by Magdaline-A — APPROVED** [review]

**Pull request merged**